MM-48567 Made registerChannelIntroButtonAction usable by all plugins

hmhealey opened the pull request:
The main reason for this is that the existing code for this only supports the Focalboard plugin using its old `focalboard` ID which isn't compatible with the `boards` ID that it uses in product mode. At the same time, none of this seemed like it was inherently Boards-specific, so I renamed some things and added support for multiple buttons being added through this plugin API.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48567

#### Related Pull Requests
https://github.com/mattermost/mattermost-webapp/pull/11800
https://github.com/mattermost/mattermost-webapp/pull/11801
https://github.com/mattermost/mattermost-webapp/pull/11802

#### Release Note
```release-note
- Made registerChannelIntroButtonAction plugin API usable by plugins other than Boards
- Fixed Create a Board button in channel intro not appearing when running Boards without a plugin
```

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

hmhealey commented on components/post_view/channel_intro_message/pluggable_intro_buttons/pluggable_intro_buttons.tsx:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    const buttons = props.pluginButtons.map((buttonProps) => {
+        // text defaults to "Create a board" because the original version of this API had hardcoded text and was
+        // built specifically for Boards
+        const text = buttonProps.text ?? intl.formatMessage({id: 'intro_messages.createBoard', defaultMessage: 'Create a board'});
```
Boards doesn't pass in any text only "Boards" for a `tooltipText` option that's not used, so this is a bit of a hack to ensure that this is backwards compatible with the plugin. We can remove it once we stop supporting the Boards plugin

hmhealey commented on plugins/registry.ts:
```diff
@@ -187,20 +187,19 @@ export default class PluginRegistry {
     // - icon - React element to use as the button's icon
-    // - action - a function called when the button is clicked, passed the channel and channel member as arguments
-    // - tooltipText - string or React element shown for tooltip appear on hover
+    // - action - a function called when the button is clicked
+    // - text - a localized string to use as the button's text
```
@sbishel We never actually passed any arguments into `action` before, so I just removed the mention it from this comment. If you'd prefer, I can add those back in and have the arguments passed in though like we previously wanted to do.

I also removed `tooltipText` since we never actually used it

sbishel replied:
Yeah, I think we ended up using the "currentChannel" and therefore didn't need the arguments passed. Other plugins may need/want arguments, but from boards side, we don't need them.

hmhealey replied:
This is no longer needed since I just renamed `tooltipText` to `text` instead of adding a second field

calebroseland commented on plugins/registry.ts:
```diff
@@ -187,20 +187,19 @@ export default class PluginRegistry {
+    // - text - a localized string to use as the button's text
     registerChannelIntroButtonAction = reArg([
         'icon',
         'action',
```
Without this, `'Boards'` in https://github.com/mattermost/focalboard/blob/eceb406535b80cd645678b9273c1ace5c62f689e/mattermost-plugin/webapp/src/index.tsx#L332 won't get mapped to `text` in the destructured param. (Unless it was switched to the object arg signature.)

```suggestion
        'action',
        'text',
```

hmhealey replied:
Yep, that's expected. The text doesn't match ("Create a Board" vs "Boards"), and I didn't want to have to break that, so I just chose not to add `text` as an option if you're using the old API.

Personally, I'd be a fan of deprecating the non-object way of calling these at some point as well, so I don't think it's as important for that to support all the options. I haven't actually ran that by anyone yet though

calebroseland replied:
Actually, we need to keep 'text' in the mapping for the time being. `reArg` has a keys check to verify the mapping is supposed to be done or not https://github.com/mattermost/mattermost-webapp/blob/d2e60742b5cbd342981e542cbeb31eb802063d6e/utils/func.ts#L30


Without consistent keys between `keyOrder` and `TArgs`, when calling via the new object-arg signature, `reArg` treats the object provided as `icon` (first argument), and this causes a hard crash.


I'm all for the deprecating route, but `reArg` will have to be improved a bit before we can do hard cutoffs like this.

hmhealey replied:
Yeah, I've wanted to do something similar, but with a new package instead of adding it to `@mattermost/types`. I've been preferring the idea of more smaller packages lately that are more focused on use case rather than "here's all  this sort of code" like how most of our code base is organized now. There's definitely a potential that that just means everyone needs to add 12 `@mattermost/` dependencies at the start of a project that I am a bit worried about though.

Good catch. I'll re-add the `text` parameter but mark that it's unused then

hmhealey commented on plugins/registry.ts:
```diff
@@ -187,20 +187,19 @@ export default class PluginRegistry {
         icon: ReactResolvable;
         action: PluginComponent['action'];
-        tooltipText: ReactResolvable;
+        text: string;
```
I went with `text` being a plain string instead of a React component since that meant we don't need to pass a separate aria-label. On one hand, that means we have less props to pass around, but it does mean we lose some options around styling and formatting localized text

hmhealey commented on plugins/registry.ts:
```diff
@@ -218,12 +217,6 @@ export default class PluginRegistry {
 
-        store.dispatch({
-            type: ActionTypes.RECEIVED_PLUGIN_COMPONENT,
-            name: 'MobileChannelIntroButton',
```
There were no other references to this name anywhere, so I think it was just copy and pasted from the `ChannelHeader` button

calebroseland commented on components/post_view/channel_intro_message/pluggable_intro_buttons/pluggable_intro_buttons.tsx:
```diff
@@ -0,0 +1,46 @@
+            <button
+                key={buttonProps.id}
+                className={'intro-links color--link channelIntroButton style--none'}
+                onClick={buttonProps.action}
```
Should we consider providing channel context specifics to reduce the need to go select these manually? (like in `registerCallButtonAction`, the action there has a signature of `(currentChannel: Channel, myCurrentChannelMembership: ChannelMembership) => void)`

hmhealey replied:
Yeah, we probably should. I removed the comment about that from the registry method because we weren't actually passing them, but we probably need to to make this at all useful. I also remember looking at how the Boards code that uses this works, and it immediately reads Redux to get that information which is something we'd definitely like to reduce the need for

calebroseland reviewed (approved):
Looks good to me. Useful addition!

sbishel reviewed (commented):
Thanks for doing this. 

> Yep, that's expected. The text doesn't match ("Create a Board" vs "Boards"), and I didn't want to have to break that, so I just chose not to add text as an option if you're using the old API.

I think this [PR](https://github.com/mattermost/focalboard/pull/4377) would make things easier and cleaner. Then you can just use the text as the button title and get rid of all the "boards" workarounds/hacks. I'm happy to get it merged.

hmhealey commented:
As Scott and I discussed in his Focalboard PR, I removed the special handling for if `text` was empty and removed the redundant option to `registerChannelIntroButtonAction` (there's only one `text` option instead of that and `tooltipText`).

This does mean that anyone using an old version of the Boards plugin with the web app would see that button with the text "Boards" instead of "Create a Board", but that's a minor bug. It's also unlikely to happen because this is the release where we're turning on product mode by default, and even if we weren't, Boards is prepackaged so people should always be on the latest anyway

sbishel reviewed (approved):
Awesome. Looks good.
